File: MainApp/src/main/java/com/sleticalboy/weight/xrefresh/XRefreshView.java
package com.sleticalboy.weight.xrefresh;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.sleticalboy.learning.R;

/**
 * Created on 18-2-3.
 *
 * @author leebin
 * @version 1.0
 * @description
 */
public class XRefreshView extends FrameLayout {

    private static final String TAG = "XRefreshView";

    private ViewGroup mEmptyView;
    private ViewGroup mErrorView;
    private ViewGroup mProgressView;
    private int mEmptyId;
    private int mProgressId;
    private int mErrorId;
    private Context mContext;

    public XRefreshView(@NonNull Context context) {
        this(context, null);
    }

    public XRefreshView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public XRefreshView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mContext = context;
        init(attrs);
    }

    private void init(AttributeSet attrs) {
        initView();
        initAttrs(attrs);
    }

    private void initView() {
        if (isInEditMode()) {
            return;
        }
    }

    private void initAttrs(AttributeSet attrs) {
        if (attrs == null) {
            return;
        }
        TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.XRefreshView);
        try {
            mEmptyId = a.getResourceId(R.styleable.XRefreshView_layout_empty, 0);
            mErrorId = a.getResourceId(R.styleable.XRefreshView_layout_error, 0);
            mProgressId = a.getResourceId(R.styleable.XRefreshView_layout_progress, 0);
        } finally {
            a.recycle();
        }
    }

    public void setEmptyView(View emptyView) {
        mEmptyView.removeAllViews();
        mEmptyView.addView(emptyView);
    }

    public void setEmptyView(int emptyView) {
        mEmptyView.removeAllViews();
        View.inflate(mContext, emptyView, mEmptyView);
    }

    public void setErrorView(View errorView) {
        mErrorView.removeAllViews();
        mErrorView.addView(errorView);
    }

    public void setErrorView(int errorView) {
        mErrorView.removeAllViews();
        View.inflate(mContext, errorView, mErrorView);
    }

    public void setProgressView(View progressView) {
        mProgressView.removeAllViews();
        mProgressView.addView(progressView);
    }

    public void setProgressView(int progressView) {
        mProgressView.removeAllViews();
        View.inflate(mContext, progressView, mProgressView);
    }
}

File: MainApp/src/main/java/com/sleticalboy/weight/xrefresh/XRefreshView.kt
package com.binlee.weight.xrefresh

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.binlee.learning.R

/**
 * Created on 18-2-3.
 *
 * @author leebin
 * @version 1.0
 * @description
 */
class XRefreshView @JvmOverloads constructor(
  private val mContext: Context,
  attrs: AttributeSet? = null,
  defStyleAttr: Int = 0
) : FrameLayout(mContext, attrs, defStyleAttr) {
  private val mEmptyView: ViewGroup? = null
  private val mErrorView: ViewGroup? = null
  private val mProgressView: ViewGroup? = null
  private var mEmptyId = 0
  private var mProgressId = 0
  private var mErrorId = 0
  private fun init(attrs: AttributeSet?) {
    initView()
    initAttrs(attrs)
  }

  private fun initView() {
    if (isInEditMode) {
      return
    }
  }

  private fun initAttrs(attrs: AttributeSet?) {
    if (attrs == null) {
      return
    }
    val a = context.obtainStyledAttributes(attrs, R.styleable.XRefreshView)
    try {
      mEmptyId = a.getResourceId(R.styleable.XRefreshView_layout_empty, 0)
      mErrorId = a.getResourceId(R.styleable.XRefreshView_layout_error, 0)
      mProgressId = a.getResourceId(R.styleable.XRefreshView_layout_progress, 0)
    } finally {
      a.recycle()
    }
  }

  fun setEmptyView(emptyView: View?) {
    mEmptyView!!.removeAllViews()
    mEmptyView.addView(emptyView)
  }

  fun setEmptyView(emptyView: Int) {
    mEmptyView!!.removeAllViews()
    inflate(mContext, emptyView, mEmptyView)
  }

  fun setErrorView(errorView: View?) {
    mErrorView!!.removeAllViews()
    mErrorView.addView(errorView)
  }

  fun setErrorView(errorView: Int) {
    mErrorView!!.removeAllViews()
    inflate(mContext, errorView, mErrorView)
  }

  fun setProgressView(progressView: View?) {
    mProgressView!!.removeAllViews()
    mProgressView.addView(progressView)
  }

  fun setProgressView(progressView: Int) {
    mProgressView!!.removeAllViews()
    inflate(mContext, progressView, mProgressView)
  }

  companion object {
    private const val TAG = "XRefreshView"
  }

  init {
    init(attrs)
  }
}